File: Recursion/MergeSort.java
package Recursion;

import java.util.Arrays;

public class MergeSort {

	public static void main(String[] args) {
		int[] a = {10,9,8,7,6,5,4,3,2,1};
		int si=0;
		int ei = a.length-1;
		int[] arr = mergeSort(a,si,ei);
		for(int i=0;i<a.length;i++) {
			System.out.println(arr[i]);
		}
	}
	
	public static int[] mergeSort(int[]a, int si, int ei) {
		
		if(si>=ei) {
			
			return a;
		}
			
		int mid = (si+ei)/2;
		

         mergeSort(a,si,mid);

		 mergeSort(a,mid+1,ei);

		return merge(a,si,ei,mid);

	}
	
	public static int[] merge(int[]a, int si, int ei, int mid) {
		int[] l = new int[mid-si+1];
		int[] r = new int[ei-mid];
		
		int l1=0;
		for(int i=si;i<=mid;i++) {
			l[l1] = a[i];
			l1++;
			
		}
		
		int r1=0;
		for(int i=mid+1;i<=ei;i++) {
			r[r1] = a[i];
			r1++;
		}
		
		int l2=0,r2=0,k=si;
		while(l2<=l.length-1 && r2<=r.length-1) {
			if(l[l2]<r[r2]) {
				a[k]=l[l2];
				k++;
				l2++;
			}
			else {
				a[k]=r[r2];
				k++;
				r2++;
			}
		}
		
		while(l2<=l.length-1) {
			a[k]=l[l2];
			k++;
			l2++;
		}
		while(r2<=r.length-1) {
			a[k]=r[r2];
			k++;
			r2++;
		}
		return a;
	}	
}

File: Arrays/MergeSort.cpp
#include <iostream>
#include <map>
#include<algorithm>
#include<bits/stdc++.h>
using namespace std;

void merge(int a[], int start, int end, int mid){

    int ps1 = start, pe1 = mid; // start and end of the left sub arr.

    int size1 = pe1 - ps1 +1; // size of ls

    int ps2 = mid+1, pe2 = end; // start and end of the right sub arr.

    int size2 = pe2 - ps2 +1; // size of rs

    int index = start;

    int ls[size1];
    int rs[size2];

    for(int i=0;i<size1;i++){
        ls[i] = a[ps1+i];
    }

    for(int i=0;i<size2;i++){
        rs[i] = a[ps2+i];
    }

    int p1 = 0, p2 = 0;
    int k = ps1;
    while(p1<size1&&p2<size2){

        if(ls[p1]<rs[p2]){
            a[k] = ls[p1];
            p1++;
            k++;
        }else{
            a[k] = rs[p2];
            p2++;
            k++;
        }
    }

    if(p1<size1){
        while(p1<size1){
            a[k] = ls[p1];
            k++;
            p1++;
        }
    }

    if(p2<size2){
        while(p2<size2){
            a[k] = rs[p2];
            k++;
            p2++;
        }
    }

}

void mergeSort(int a[], int start, int end){

    if(start == end){
        return;
    }

    int mid = (start + end)/2;

    mergeSort(a, start, mid);
    mergeSort(a, mid+1, end);
    merge(a,start,end,mid);

}

int main() {

    int a[7] = {5,4,3,2,1,1,1};
    mergeSort(a,0,6); 

    for(int el:a){
        cout<<el<<endl;
    }
}
